Add filter option

maximumhallinan opened the pull request:
In response to your most recent comment on [issue #10](https://github.com/sindresorhus/strip-css-comments/issues/10#issuecomment-118329158), this PR adds support for a filter option. The filter option is a function that "gets the comment contents and returns a boolean whether to strip the comment". If true, the comment is stripped. If false, the comment is preserved. This PR includes relevant tests and documentation.

maximumhallinan commented:
I just noticed there's an indentation issue on lines 10 and 11 that doesn't appear in my text editor. If this PR is good otherwise, I'll fix and resubmit.

sindresorhus commented on index.js:
```diff
@@ -6,6 +6,11 @@ module.exports = function (str, opts) {
 	var currentChar = '';
 	var insideString = false;
+
+	var filter = !!(opts.filter && opts.filter.constructor && opts.filter.call && opts.filter.apply);
```
```js
var filter = typeof opts.filter === 'function';
```

sindresorhus commented on index.js:
```diff
@@ -27,8 +32,19 @@ module.exports = function (str, opts) {
 				// start skipping until we reach end of comment
 				for (var j = i + 2; j < str.length; j++) {
 					if (str[j] + str[j + 1] === '*/') {
+						if (filtering){
```
`if (filtering) {`

sindresorhus commented on index.js:
```diff
@@ -36,7 +52,7 @@ module.exports = function (str, opts) {
 		}
 
-		ret += currentChar;
+		if (!filtering){ret += currentChar;}
```
```js
if (!filtering) {
    ret += currentChar;
}
```

sindresorhus commented:
Needs a test with `preserve` and `filter` set at once.

maximumhallinan commented:
Thanks for the feedback, Sindre! 

Good point about testing the `preserve` and `filter` combination. What's the desired behavior? At the moment, all comments matched by the filter function are stripped regardless of the value of `opts.all`. If `opts.all !== true`, should filter exclude important comments? 

If you're going to merge, would you want this all re-submitted as a single commit?

sindresorhus commented:
> At the moment, all comments matched by the filter function are stripped regardless of the value of opts.all

Makes sense like it is. `opts.filter` just overrides `opts.all`.

> If you're going to merge, would you want this all re-submitted as a single commit?

Nah, I can squash when I merge.

maximumhallinan commented:
👍

maximumhallinan commented:
Hey @sindresorhus , sorry to bug you about this. Just want to make sure that it's clear that the last commit included the tests you asked for. Since I followed that with a question, it occurs to me that might not have been clear.

sindresorhus commented:
Yeah, been meaning to finish reviewing this, but summer.

I've also been a bit conflicted about the API. What do you think about merging `all` and `filter` into a `preserve` option?

The `preserve` option would accept either a regex, function or false of whether to preserve the comment. The default value would simply be `/^!/`. I kinda feel this would make the API cleaner. And allowing a regex would be faster and solve the common case for most, but also have a function available for more advanced dynamic cases.

-

*(Sorry about the churn, but I'm trying to make the API as good as possible now. Harder to do that later).*

sindresorhus commented:
// @kevva @arthurvr @floatdrop

maximumhallinan commented:
The overlap of `all` and `filter` feels messy to me too. I'm glad to have the opportunity to step back and reconsider the approach.

My bias is towards simplicity. For that reason, I like your idea of a single config option for all preservation-related functionality. Because I prefer simplicity, I would limit `preserve` to accept a function that takes the comment as an argument and returns `true` to preserve and `false` to strip. If the option is undefined or not a function, no comments are preserved.

On excluding regex:

It's not hard for a user to wrap a regex in a function. The function provides maximum flexibility for complex cases and can be used with ease to address the simpler cases without requiring the internal complexity of checking for regex vs. function and handling each type correctly.

On excluding `false`:

I'm skeptical of the value of privileging any syntax (such as `/*! */`) as a default comment to preserve. I can't say why exactly except that it doesn't feel necessary. And again, the additional type checking, such as checking for a boolean that overrides this default, seems like unwelcome internal complexity.

Overall, I sense a desire to address varying levels of user engagement or sophistication. Rather than building complexity into the API, it might be better to offer examples in the documentation that could be used as a model for the simpler cases and assume that a user with a complex case has the sophistication to write the function required to handle it.

sindresorhus commented:
I'm also strongly biased towards simplicity, but not just the implmentation. The API has to be simple too. While I agree on your point of excluding `false` and dropping the `/* */` default, I still think supporting regex is a nice touch. A function is nice for ultimate flexibility, but I think a regex can cover most needs. I would maybe agree with you if arrow functions were mainstream, but right now a regex is nicer and shorter (and also faster than a function call, but don't have any numbers to back that up).

The logic can be simple. If undefined don't do anything, if defined and a function, call function, if defined and not a function, call the regex. Not a lot of added complexity IMHO.

Qix- commented:
- `preserve: false` - don't preserve important comments
- `preserve: true` (default) - preserve important comments
- `preserve: function(text) { ... }` - true/false return whether or not to keep it
- `preserve: RegExp` - `RegExp.prototype.test.bind(regex)`

Seems clean enough to me. Two properties would be confusing, and aren't interoperable when used together.

sindresorhus commented:
Is it expected to preserve "important" comments by default? That's my question. If it's a super common case I think it *does* make sense to default to that, as it will require less config for most people.

See https://github.com/sindresorhus/strip-css-comments/pull/8 where it was added.

// @7studio @kevva @arthurvr @floatdrop

maximumhallinan commented:
@sindresorhus you make an excellent point about simplicity being a desired quality of the API as much as the implementation. I'm on board with regex. 

I'm also happy to retain support for preserving `/*!` by default. After reading the original [request](https://github.com/sindresorhus/strip-css-comments/issues/4), I'm more sympathetic though still resistant to a feature which anticipates a convention external to the module itself. But as with regex, you've made excellent points in favor of maintaining support for this feature. IMHO, if we're going to address `/*!` at all, it is most clean to preserve that syntax by default. Is there a reason not to set that behavior as the default?

Qix- commented:
Well I don't see a lot of comments that start with `/*!`. My guess is that if I came across it in the wild, my first instinct would be to assume it's special in some way. I don't see why you wouldn't keep it in there.

sindresorhus commented:
Ok, let's go with function/regex and `/*!` preserved by default. With this it should be easy to still support the old `all` option too, so we don't end up breaking anything :)

Qix- commented:
:+1:

maximumhallinan commented:
Sounds good. New pull request?

sindresorhus commented:
Nah, just push commits to this PR. Better to have the discussion in one place.

maximumhallinan commented:
I've updated the feature branch with support for `opts.preserve` and removed support for `opts.filter`. The API is:
- `preserve: true` &mdash; (default) preserve comments that use the `/*! */` syntax;
- `preserve: false` &mdash; strip all comments;
- `preserve: [RegExp]` &mdash; preserve comments that match a regular expression. The comment text but not the comment syntax (`/**/`) will be tested by the RegExp.
- `preserve: function (comment) { ... }` &mdash; a function that returns `true` to preserve the comment or `false` to strip it. The comment is invoked with a single argument: the string found between the comment syntax, `/**/`.

`opts.all: true` is supported for the CLI and backwards compatibility.

Qix- commented:
Should the string passed to the function/regex have the `/**/` in them? I'd be inclined to say no.

sindresorhus commented:
No

sindresorhus commented on index.js:
```diff
@@ -3,11 +3,26 @@ module.exports = function (str, opts) {
 
+	//use preserveFilter as an interface for opts.preserve function and RegExp values
+	if (opts !== undefined && opts.preserve !== undefined) {
+		if ((typeof opts.preserve).toLowerCase() === 'function') {
```
lowercasing this is moot

maximumhallinan replied:
True. Time for me to dispense with my paranoia that someday `typeof` will return a string in a case other than lower case.

sindresorhus commented on index.js:
```diff
@@ -3,11 +3,26 @@ module.exports = function (str, opts) {
+		if ((typeof opts.preserve).toLowerCase() === 'function') {
+			preserveImportant = false;
+			preserveFilter = opts.preserve;
+		} else if ((opts.preserve.constructor.name).toLowerCase() === 'regexp') {
```
the `if` is moot as we can just assume it's an regex.

Qix- replied:
For the record, interpreters usually optimize `typeof x === y` statements. I've yet to see one optimize `x.constructor.name === y` statements. `instanceof` is going to be faster almost always, unless you're testing for a primitive type other than function, which should indeed be `x.constructor.name === y`.

maximumhallinan replied:
@Qix- , thanks for the recommendation! I wasn't aware of the performance consideration.

@sindresorhus, please see below about assuming `RegExp`.

sindresorhus commented on index.js:
```diff
@@ -21,16 +36,29 @@ module.exports = function (str, opts) {
+					}
 				}
-				// skip i to the end of the comment
+				//resume iteration over CSS string from the end of the comment
```
space after `//`

maximumhallinan replied:
I've added a space to each comment.

sindresorhus commented on index.js:
```diff
@@ -3,11 +3,26 @@ module.exports = function (str, opts) {
 	var currentChar = '';
 	var insideString = false;
+	var preserveFilter;
+	var preserveImportant = (opts.preserve === false || opts.all === true) ? false : true;
```
This can be simplified to remove the ` ? false : true`.

maximumhallinan replied:
Simplified. Thanks for the nudge.

sindresorhus commented on index.js:
```diff
@@ -3,11 +3,26 @@ module.exports = function (str, opts) {
 	var ret = '';
 
+	//use preserveFilter as an interface for opts.preserve function and RegExp values
+	if (opts !== undefined && opts.preserve !== undefined) {
```
`opts !== undefined` is moot as `opts` will never be undefined.

maximumhallinan replied:
Thanks! I overlooked the conditional assignment.

sindresorhus commented on index.js:
```diff
@@ -3,11 +3,26 @@ module.exports = function (str, opts) {
+				return opts.preserve.test(comment);
+			};
+		}
+	}
```
Could maybe be simplified to:

```js
if (opts.preserve !== undefined) {
	preserveImportant = false;
	preserveFilter = typeof opts.preserve === 'function' ?
		opts.preserve : function (comment) {
		return opts.preserve.test(comment);
	};
}
```

maximumhallinan replied:
I hope I've understood correctly what you are suggesting. Apologies if I haven't.

I'm not assuming that `opts.preserve` is either a function or a `RegExp` (and nothing else) because it could also be a boolean. The outer conditional is satisfied if `opts.preserve === true` or `opts.preserve === false`. If `opts.preserve` is a boolean, `preserveFilter` should be `undefined`. 

Also, by explicitly checking for both a RegExp and a function, calling strip-css-comments with an invalid `opts.preserve` value such as `'false'` won't have a catastrophic effect. Instead, invalid configuration is ignored. I'm not sure if this is more desirable than throwing a validation error. 

To simplify, I've removed `if (opts.preserve !== undefined)` and am now just checking the type of `opts.preserve`.

sindresorhus replied:
Alright, but the regex check is a bit fragile. You can use: https://github.com/sindresorhus/is-regexp

maximumhallinan replied:
Added is-regexp as a dependency.

sindresorhus commented on readme.md:
```diff
@@ -18,11 +18,28 @@ var stripCssComments = require('strip-css-comments');
+//assign the preserve option a regular expression to strip comments not matching the pattern
+stripCssComments(
+	'/*! <copyright> */ body { /* unicorns */color: hotpink; }', 
+	{preserve: /^\\!/}
```
`{preserve: /^\\!/}` => `{preserve: /^\!/}`

maximumhallinan replied:
Sorry about this. My Markdown preview uses JavaScript and was requiring me to add an extra escape character. I meant to remove them.

sindresorhus commented on readme.md:
```diff
@@ -35,14 +52,18 @@ Type: `string`
+Default: `true`
 
-Whether *important* CSS comments *(those starting with `/*!`)* should be stripped.
+- `preserve: true` &mdash; (default) preserve comments that use the `/*! */` syntax;
```
Just use a normal `-`

maximumhallinan replied:
Replaced. Out of curiosity, what's the reason? I was taught that an em-dash is more typographically/grammatically correct in this case.

sindresorhus replied:
I guess it might be more grammatically correct, but I prefer the look of the normal dash.

https://en.wikipedia.org/wiki/Dash#Em_dash

You also shouldn't use spaces around an em-dash.

Qix- commented on test.js:
```diff
@@ -20,14 +20,50 @@ test(function (t) {
+        strip('body{/*!##foo##*//*foo*/}', {
+                preserve: function(comment){return /^##foo##/.test(comment);}
+        }) === 'body{}');
+
```
Please check your editor indentation :)

maximumhallinan replied:
Thanks!

Qix- commented on index.js:
```diff
@@ -21,16 +36,29 @@ module.exports = function (str, opts) {
 					if (str[j] === '*' && str[j + 1] === '/') {
+						if (preserveFilter) {
+							//evaluate comment text
+							ret = preserveFilter(comment) ? ret += ('/*' + comment + '*/') : ret;
```
True ternary result should just be `ret + ...` instead of `ret += ...` because you're assigning to `ret` anyway.

maximumhallinan replied:
Good point. Thanks!

Qix- commented on readme.md:
```diff
@@ -35,14 +52,18 @@ Type: `string`
+- `preserve: [RegExp]` &mdash; preserve comments that match a regular expression. The comment text but not the comment syntax (`/**/`) will be tested by the RegExp.
+- `preserve: function (comment) { ... }` &mdash; a function that returns `true` to preserve the comment or `false` to strip it. The comment is invoked with a single argument, the string found between the comment syntax, `/**/`.
 
 
```
Remove this line

arthurvr replied:
@Qix- Why? It's unrelated to this PR and 2 lines between sections of the readme is consistent. That's the way we do it everywhere.

arthurvr replied:
Why? These are consistent, unrelated to this PR and there because the author chose to do so.

Qix- replied:
Saw the other ones, didn't realize they were already there. Disregard.

arthurvr commented on readme.md:
````diff
@@ -66,14 +87,12 @@ $ strip-css-comments --help
 $ npm run bench
 ```
 
-
````
Why did you remove these lines? These are consistent and there because the author chose to do so. Don't change unrelated things in this PR ;)

maximumhallinan replied:
You're right. Thanks! 

I've returned these spaces and after comparing `readme.md` on this branch with the master branch, I've deleted one line and added another.

Qix- commented on readme.md:
```diff
@@ -54,16 +75,16 @@ $ npm install --global strip-css-comments
-    $ strip-css-comments < src/app.css --all
+	Example
+		$ strip-css-comments src/app.css > dist/app.css
+		$ strip-css-comments < src/app.css --all
```
Why are these edited?

maximumhallinan replied:
Not sure. I didn't edit them directly. Perhaps something to do with space to tab conversion. I'm using the editorconfig plugin so I'm surprised to have formatting issues. I went through the readme again and made a few other formatting corrections.

Qix- commented:
LGTM other than that other nit.

sindresorhus commented:
Sorry about the delay. I'm travelling atm. Landed!

Thanks for the hard work on this @maximumhallinan :)

Available in 2.1.0.

maximumhallinan commented:
This was fun to go back and forth on. Thanks for the merge!

@Qix, @arthurvr thanks for the reviews!
